crates/ra_assists/src/handlers/reorder_fields.rs

```diff
@@ -0,0 +1,207 @@
+use std::collections::HashMap;
+
+use itertools::Itertools;
+
+use hir::{Adt, ModuleDef, PathResolution, Semantics, Struct};
+use ra_ide_db::RootDatabase;
+use ra_syntax::ast::{Name, Pat};
+use ra_syntax::{
+    ast,
+    ast::{Path, RecordField, RecordLit, RecordPat},
+    AstNode,
+};
+
+use crate::{
+    assist_ctx::{Assist, AssistCtx},
+    AssistId,
+};
+
+pub(crate) fn reorder_fields(ctx: AssistCtx) -> Option<Assist> {
+    reorder_struct(ctx.clone()).or_else(|| reorder_struct_pat(ctx))
+}
+
+fn reorder_struct(ctx: AssistCtx) -> Option<Assist> {
+    let record: RecordLit = ctx.find_node_at_offset()?;
+    reorder(ctx, &record, &record.path()?, field_name)
+}
+
+fn field_name(r: &RecordField) -> String {
+    r.name_ref()
+        .map(|name| name.syntax().text())
+        .unwrap_or_else(|| r.expr().unwrap().syntax().text())
+        .to_string()
+}
+
+fn reorder_struct_pat(ctx: AssistCtx) -> Option<Assist> {
+    let record: RecordPat = ctx.find_node_at_offset()?;
+    reorder(ctx, &record, &record.path()?, field_pat_name)
+}
+
+fn field_pat_name(field: &Pat) -> String {
+    field.syntax().children().find_map(Name::cast).map(|n| n.to_string()).unwrap_or_default()
+}
+
+fn reorder<R: AstNode, F: AstNode + Eq + Clone>(
+    ctx: AssistCtx,
+    record: &R,
+    path: &Path,
+    field_name: fn(&F) -> String,
+) -> Option<Assist> {
+    let ranks = compute_fields_ranks(path, &ctx)?;
+    let fields: Vec<F> = get_fields(record);
+    let sorted_fields: Vec<F> =
+        sort_by_rank(&fields, |f| *ranks.get(&field_name(f)).unwrap_or(&usize::max_value()));
+
+    if sorted_fields == fields {
+        return None;
+    }
+
+    ctx.add_assist(AssistId("reorder_fields"), "Reorder record fields", |edit| {
+        for (old, new) in fields.into_iter().zip(sorted_fields) {
+            edit.replace_ast(old, new);
+        }
+        edit.target(record.syntax().text_range())
+    })
+}
+
+fn get_fields<R: AstNode, F: AstNode>(record: &R) -> Vec<F> {
+    record.syntax().children().flat_map(|n1| n1.children()).filter_map(|n3| F::cast(n3)).collect()
+}
+
+fn sort_by_rank<F: AstNode + Clone>(fields: &[F], get_rank: impl FnMut(&F) -> usize) -> Vec<F> {
+    fields.iter().cloned().sorted_by_key(get_rank).collect()
+}
+
+fn struct_definition(path: &ast::Path, sema: &Semantics<RootDatabase>) -> Option<Struct> {
+    match sema.resolve_path(path) {
```

Comment:
Ideally this should use `variant_resolutions` and not just resolve_paths, but it's not clear how to best expose this API for an IDE, so I guess this is ok as an initial impl